[Sign-in Revamp]: New Auth Modal + Updated Wallet behaviours

mzparacha opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->

## Link to Issue
Closes: https://github.com/hicommonwealth/commonwealth/issues/6384

## Description of Changes
- Adds new Auth Modal UI wth responsive screen behaviours
- Integrated Cosmos/Substrate and Solana wallets in the new Modal i.e Polkadot, Phantom, Leap, and Keplr. EVM wallets will be integrated in https://github.com/hicommonwealth/commonwealth/issues/6387
- Integrated SSO options in the new modal i.e Google, X (Twitter), Discord, and Github. Email login will be integrated in https://github.com/hicommonwealth/commonwealth/issues/6386
- Updated behaviours for wallet options visibility.
  - EVM wallet options (+ WalletConnect) will now only show in EVM communities, in all non-community pages and as requested by the `showWalletsFor` in the new modal
  - Cosmos wallet options will now only show in Cosmos communities, in all non-community pages and as requested by the `showWalletsFor` in the new modal.
  - Substrate wallet options will now only show in Substrate communities, in all non-community pages and as requested by the `showWalletsFor` in the new modal
  - Solana wallet options will now only show in Solana communities, in all non-community pages and as requested by the `showWalletsFor` in the new modal
- Now during a new wallet account signup, we no longer require the user to input their username/profile-pic, instead we directly create an account for them with the default `Anonymous` name
- Updated all instances of the old `LoginModal` with the new `AuthModal`
- New `AuthModal` is behind the new `FLAG_NEW_SIGN_IN_MODAL` flag, to see changes enable the flag in `.env`

## "How We Fixed It"
N/A

## Test Plan
- Enable `FLAG_NEW_SIGN_IN_MODAL` in `.env`
- Install and setup `Keplr`, `Leap`, `Phantom` and `Polkadot` wallets if you don't have already.
- Logout (if logged in) and go to any non-community page, open the sign-in modal and verify you see all available wallets for all 4 supported chains (EVM, Cosmos, Solana, Substrate) in the wallets tab -> `Wallet Connect`, `Keplr` and `Leap`, `Phantom` and `Polkadot`, verify that signing in with all wallets (excluding wallet connect) works in those pages. Also verify that signing in with all SSO's works in those pages
- Logout (if logged in) and go to any Solana based community (ex: `/solana/discussions`), open the sign-in modal and verify you see only the `Phantom` wallet in the wallets tab, verify that signing in `Phantom` wallet and all `SSO's` works in those pages.
- Logout (if logged in) and go to any Substrate based community (ex: `/edgeware/discussions`), open the sign-in modal and verify you see only the `Polkadot` wallet  in the wallets tab, verify that signing in `Polkadot` wallet and all `SSO's` works in those pages.
- Logout (if logged in) and go to any Ethereum based community (ex: `/dydx/discussions`), open the sign-in modal and verify you see only the `WalletConnect` option  in the wallets tab, verify that signing in with all `SSO's` works in those pages. (WalletConnect will be integrated later)
- Logout (if logged in) and go to new community page (`/createCommunity`), click on any option i.e `Ethereum`, `Cosmos` or `Polygon` and verify that `Ethereum` + `Polygon` when clicked opens the `AuthModal` with only the `WalletConnect` option in the wallets tab, and `Cosmos` option when clicked opens the `AuthModal` with only the `Leap` and `Keplr` options in the wallets tab. All SSO's will be available in the SSO's tab. Verify that signing-in with different wallets works in this page.

## Deployment Plan
Part of https://github.com/hicommonwealth/commonwealth/milestone/38

## Other Considerations
`FLAG_NEW_SIGN_IN_MODAL` should not be enabled in prod yet.

masvelio reviewed (commented):
1. Tried logging in with keplr on landing page but nothing happens. When I reloaded a page it seemed to work.

https://github.com/hicommonwealth/commonwealth/assets/14819225/040bebca-8c8a-4d32-b17c-8655131931c5

2. X button is misplaced vs design

![image](https://github.com/hicommonwealth/commonwealth/assets/14819225/c377e6d4-0d01-413d-90e6-d4dbddbe5be6)

masvelio commented on packages/commonwealth/client/scripts/views/modals/AuthModal/AuthModal.tsx:
```diff
@@ -0,0 +1,160 @@
+  const tabsList = [
+    {
+      name: 'Wallet',
+      options: [
+        // Note: If `showWalletsFor` is present then show wallets for that chain, else
+        // only show ethereum based wallets options on non-community pages or on ethereum based communities
+        ...((!showWalletsFor
+          ? !app?.chain?.base
+          : showWalletsFor === ChainBase.Ethereum) ||
+        app?.chain?.base === ChainBase.Ethereum
+          ? ['walletConnect']
+          : []),
+        // only show cosmos based wallets on non-community pages or on cosmos based communities
+        ...(wallets || [])
+          .filter(
+            (wallet) =>
+              ((!showWalletsFor
+                ? !app?.chain?.base
+                : showWalletsFor === ChainBase.CosmosSDK) ||
+                app?.chain?.base === ChainBase.CosmosSDK) &&
+              wallet.chain === ChainBase.CosmosSDK,
+          )
+          .map((wallet) => wallet.name),
+        // only show solana based wallets on solana based communities
+        ...(wallets || [])
+          .filter(
+            (wallet) =>
+              app?.chain?.base === ChainBase.Solana &&
+              wallet.chain === ChainBase.Solana,
+          )
+          .map((wallet) => wallet.name),
+        // only show substrate based wallets on substrate based communities
+        ...(wallets || [])
+          .filter(
+            (wallet) =>
+              app?.chain?.base === ChainBase.Substrate &&
+              wallet.chain === ChainBase.Substrate,
+          )
+          .map((wallet) => wallet.name),
```
despite nice comments, it is pretty hard to read it, especially those multilines and nested ternary operators
1. How about abstract out filter function so you would pass only `ChainBase` and get filtered wallets
2. let's try to use more variables and then build up on top of them, maybe something like 
const cosmosWallets = []; const substrateWallets = [];

mzparacha replied:
done.

masvelio commented on packages/commonwealth/client/scripts/views/modals/AuthModal/AuthModal.tsx:
```diff
@@ -0,0 +1,160 @@
+    },
+    {
+      name: 'Email or Social',
+      options: ['google', 'discord', 'x', 'github', 'email'],
```
are those options typed? If not, lets either type it or use enum like thing, not just strings

mzparacha replied:
done

masvelio commented on packages/commonwealth/client/scripts/views/modals/AuthModal/AuthModal.tsx:
```diff
@@ -0,0 +1,160 @@
+      open={isOpen}
+      onClose={onClose}
+      content={
+        <section className="AuthModal">
+          <CWIcon iconName="close" onClick={onClose} className="close-btn" />
+
```
There are components that we should use for building modals like CWModalHeader, CWModalBody, CWModalFooter. Thanks to them we make sure that the paddings are the same for each modals

mzparacha replied:
I used the CWModalBody and CWModalFooter components, but wasn't able to get the proper layout/styling/variant working with CWModalHeader, it is still custom atm.

<img width="599" alt="image" src="https://github.com/hicommonwealth/commonwealth/assets/51641047/729a0202-2a1f-4191-a59d-d02af9045b72">


I think there are 2 options
1. make CWModalHeader configurable in many different ways to support the new header layout, font sizes, alignments, x-icon alignment, optional logo, label alignments, and make sure all of these combinations work consistently (this would consume time, and in favor of not breaking anything, i didn't do it right now)
2. use a custom header implementation in 1-off scenarios (like this one)

and since this is a 1-off instance of this header I believe a custom implementation works here. 

Plz lmk your thoughts, do you still think we should follow 1 here? If yes, I would be happy to create a followup and do it after the sign-in milestone.

cc: @masvelio

masvelio replied:
nope, I think #2 is good in this case. We don't want super fancy generic header that is too complicated

mzparacha replied:
Awesome, resolving this.

masvelio reviewed (commented):
Very nice job! I have several but pretty small comments. Most important is using unified CWModal and descendant components

mzparacha commented:
Thanks for the feedback @masvelio. I addressed some of your comments above.

For the one's related to Modal, i might have to update some modal components to support the styles that match this new AuthModal, like `CWModalHeader` would need to support an optional custom logo. Will do those soon.

mzparacha commented on packages/commonwealth/client/scripts/hooks/useWallets.tsx:
```diff
@@ -694,6 +699,22 @@ const useWallets = (walletProps: IuseWalletProps) => {
+        }, 1000);
+      });
+    }
+  }, [activeStep, onCreateNewAccount, onSaveProfileInfo]);
```
Fixed the race condition

mzparacha commented:
@masvelio, applied the suggestions. Some imp things if you could test

- Fixed this race condition
> Tried logging in with keplr on landing page but nothing happens. When I reloaded a page it seemed to work.

- Solana and Substrate (phantom/polkadot) wallets will now show in all community pages.


There is also 1 unresolved convo [here](https://github.com/hicommonwealth/commonwealth/pull/6520#discussion_r1478265510), would appreciate your input.

masvelio reviewed (approved):
beautiful work!

mhagel reviewed (approved):
Nice work! Tested SSO, keplr, and leap.

mzparacha commented:
Will fix failing test and merge.
